### PR TITLE
fix(core,tenancy): close tenant-isolation gaps on read paths

### LINKED ISSUE #2365: core/tenancy: tenant isolation gaps on hydration and get-by-slug; vector search ranks across tenants; collection memory shared across tenants
## Problem

Tenant isolation is enforced on collection list/get/count/select/query/junction/hierarchical/change-feed, but not on these paths:

- `collection.get('<slug>')` under an active tenant context: the interceptor rewrites any string filter to `{ id: filter, tenantId }` (`tenancy/src/interceptor.ts:338-343`) before core's UUID-vs-slug detection (`core/src/collection.ts:1420-1427`) → null on SQLite, a uuid cast error on PG. The same call without a tenant context works.
- Constructor hydration is tenant-blind: `new Model({ slug })`/`{ id }` → `loadFromSlug`/`loadFromId` call `db.get` directly (`core/src/object.ts:2310-2379`). Probed: under tenant B, `new Doc({db, slug:'secret'}).initialize()` returns tenant A's row. Writes are caught by `beforeSave`/`beforeDelete`; reads are not.
- `semanticSearch`/`findSimilarToEmbedding` compute top-K across all tenants, then filter via `list()` (`collection.ts:3455-3485`) — result starvation and a soft presence leak; the in-memory fallback loads every embedding for the class.
- `_smrt_contexts` and `_smrt_embeddings` have no tenant column; `SmrtCollection.remember()` files under `owner_id='__collection__'` (`collection.ts:2938-2941`), so a tenant-scoped collection's learned memory is shared across tenants.

## Change

- Interceptor `beforeGet`: preserve core's string-filter semantics (slug vs id) and add the tenant predicate to whichever it resolves to.
- `loadFromId`/`loadFromSlug` (and `getSavedId`) go through the same interceptor/tenant predicate as collection reads, or hydration is routed through `collection.get`.
- Vector search: pre-filter candidates by tenant (native pgvector path via a tenant column or a join; in-memory path filters before ranking).
- Collection memory: add a tenant dimension (or document that collection-level memory is shared and make `remember()` refuse under a tenant context).

## Acceptance

- probes above pass: get-by-slug works under tenant context; cross-tenant hydration returns null; semanticSearch returns only the tenant's rows and does not starve
- tests in tenancy + core

## Relationships

Independent of the schema issues (touches interceptor.ts, object.ts load paths, collection.ts search/memory). Behaviour change — schema/semantics release wave.

From the 2026-08-17 database-layer assessment (finding B2).

---
**Part of epic #2382** (database-layer gap class; report in the epic's first comment). Independent of the schema issues. Touches `interceptor.ts`, `object.ts` load paths (disjoint from #2366/#2371 regions), `collection.ts` search/memory (disjoint from #2367).

### PR DESCRIPTION
**Pull request by willgriffin**
Closes #2365 (epic #2382, finding B2).

## The gaps

Tenant isolation was enforced on collection list/get/count/select/query, but four read paths bypassed it — each probed as a real cross-tenant read or a broken read in the 2026-08-17 database-layer assessment:

1. **`collection.get('<slug>')` broke under a tenant context.** The interceptor rewrote every string filter to `{ id: filter, tenantId }` before core's UUID-vs-slug detection — null on SQLite, `invalid input syntax for type uuid` on PostgreSQL's native-uuid id column. The same call without a tenant context worked.
2. **Constructor hydration was tenant-blind.** `new Model({ slug }).initialize()` → `loadFromSlug()`/`loadFromId()`/`getSavedId()` called `db.get` directly: under tenant B they hydrated tenant A's row.
3. **`semanticSearch()` ranked across all tenants**, then filtered via `list()` — result starvation (a top-K fully occupied by another tenant's rows filtered down to `[]`) and a soft presence leak.
4. **Collection memory was shared across tenants**: `remember()`/`recall()` filed everything under one `owner_id='__collection__'` key in `_smrt_contexts`.

## The fix

- **One string-filter contract.** New `resolveGetStringFilter()` in core (`interceptors.ts`) is the single source of truth for get()'s string semantics (UUID → `{ id }`, else `{ slug, context: '' }`); `collection.get()` and the tenancy `beforeGet` both consume it, so the interceptor now adds the tenant predicate to whichever shape the string resolves to.
- **Hydration/identity through the interceptor.** `loadFromId()`, `loadFromSlug()`, `getSavedId()` and `getId()` resolve their filters through `GlobalInterceptors.executeBeforeGet` (new `interceptGetFilter()` helper; field-name keys converted to columns). `getId()` was surfaced by the adversarial review: its unrouted `(slug, context)` lookup could adopt another tenant's row id and steer a later `save()` onto the foreign row. With no interceptors registered every path is a no-op passthrough — non-tenant deployments are behavior-identical.
- **Vector search pre-filters BEFORE top-K.** `findSimilarToEmbedding()` resolves the tenant predicate through the same `beforeList` pipeline that guards collection reads (empty where in → either untouched, the injected tenant predicate, or the required-mode throw) and passes the tenant's candidate id set into `EmbeddingStorage.searchSimilar()` — chunked `object_id IN (...)` against bind-variable limits on the native vector path (per-chunk top-K merge is an exact global top-K), `Set`-filtered before ranking on the in-memory path. An explicit empty candidate set short-circuits to `[]`. The final `list({ 'id in': ... })` still re-applies the interceptor — defense in depth.
- **Collection memory keyed per tenant.** `remember`/`recall`/`recallAll`/`forget`/`forgetScope` resolve `_smrt_contexts.owner_id` to `__collection__:<tenantId>` for tenant-scoped classes under an active tenant, via the same core dispatch trust anchors as the #1782 REST read scope (`resolveDispatchTenantScope` + the triple tenant-scoped-class check). Isolation is strict — no fallback to the shared key in either direction. Non-scoped classes and no-context calls keep the shared key.

`withSystemContext()` and super-admin bypass remain the explicit, reviewed cross-tenant read paths (they short-circuit inside the interceptor itself).

## Compatibility notes

- **Required-mode hydration now fails closed**: `new Model({ id | slug }).initialize()` on a `mode: 'required'` class outside a tenant context now throws `TenantContextError` instead of silently hydrating. That is the issue's intent (reads catch up with writes, which `beforeSave` already guarded).
- Memory previously learned under the shared `__collection__` key is no longer visible to tenant-context recalls (that visibility was the leak).
- Review notes, documented in `packages/tenancy/AGENTS.md`: under super-admin bypass, reads skip filtering but memory still keys to the active tenant (tighter, not a leak); an empty-string tenant id resolves to the shared memory key (misconfiguration only — real tenant ids are uuids).
- The tenant candidate query in the vector pre-filter is deliberately unbounded — a cap would silently drop rows from ranking; the durable fix for very large tenants is a tenant column on `_smrt_embeddings` (commented at the call site).

## Out of scope / follow-up

- #2412 (filed from this review): three remaining direct `db.get()` sites — STI legacy-discriminator upgrade planning and `validateCrossPackageRefs()` (write-path adjacent, interact with #2360 / PR #2404), and the STI discriminator peek in `loadRelated()` (instance-level #1321 guard already applies).
- Upsert/natural-key tenant semantics (finding B1) are #2360 / PR #2404.

## Tests

- `packages/tenancy/src/__tests__/issue-2365-tenancy-read-gaps.test.ts` — real tenancy interceptor on in-memory SQLite, 24 tests: slug get per tenant, cross-tenant hydration returns nothing, `getSavedId`/`isSaved`/`getId` scoping, required-mode fail-closed on get and hydration, `withSystemContext` escape hatch, semanticSearch starvation/isolation/minSimilarity/empty-tenant, per-tenant memory (remember/recall/recallAll/forget) plus shared-key semantics for non-scoped classes. **10 of the original 23 fail against the unfixed code** (verified by stashing the fix); only `EmbeddingProvider.embed` is mocked (external AI call).
- `packages/core/src/__tests__/issue-2365-read-interceptor-plumbing.test.ts` — core-side seams with a plain registered interceptor (the #1782 inversion pattern): `resolveGetStringFilter` units, get/hydration/`getSavedId`/`getId` scoping incl. camelCase→column conversion, string-returning interceptors, memory keying via the dispatch hooks.
- `packages/core/src/__tests__/issue-2365-read-scope-postgres.optional.test.ts` — PostgreSQL lane: slug get under a read scope (no uuid cast error), uuid predicate bound from a JS string, hydration key conversion on native columns, candidate pre-filter with uuid id stringification.

## Validation

- `packages/tenancy` full `pnpm test` — 160 passed (12 files); one pre-existing beforeGet unit test updated to the fixed contract (it asserted the string→`{id}` rewrite that WAS the bug)
- `packages/core` full `pnpm test` — 3569 passed, 67 skipped (290 files), twice (after each commit)
- `pnpm typecheck` core + `tsc --noEmit` tenancy — clean
- `pnpm test:postgres` (core) against a disposable PostgreSQL 17 container — new #2365 suite 5/5; `change-feed-concurrency` barrier and `issue-2069` UTC failures reproduce identically on clean `db1b68d92` against the same container (pre-existing environment artifacts; CI's disposable shard is the gate)
- `npx biome check` on all touched files — clean
- `pnpm smrt dev:knowledge-check` — fresh, 0 errors / 0 warnings (after rebuild)
- `node scripts/check-standards.mjs`, `check-agents-chain.mjs` — clean (chain headroom intact)
- `node scripts/check-coverage.mjs --packages core,tenancy` — core 86.92%, tenancy 87.02% (T1 floor 80%)

## Review

Two-reviewer cycle on a3fa0c0a9 (general + adversarial tenant-isolation; Codex out of quota until Aug 19): general reviewer clean; adversarial reviewer verified memory-key injectivity, interceptor/dispatch consistency, `$N` chunk numbering, top-K merge equivalence, empty-allowlist short-circuits, the loadFromId error contract, and byte-identical non-tenant behavior — and surfaced the `getId()` gap, fixed in a5f693245 and verified clean by a focused recall review.

```json
{"schema":"hv-agent-run:v1","policy_revision":"1.0.0","runtime":"claude","session":"2e828b4a-bf86-463b-8159-3ee4e9f1275f","issue":"https://github.com/happyvertical/smrt/issues/2365"}
```

### CODE REVIEW

File: packages/core/src/object.ts [hunk 1]
```diff
@@ -1598,19 +1620,68 @@ export class SmrtObject extends SmrtClass {
...
+      interceptorContext,
+    );
+    // An interceptor may hand back a string (the public beforeGet contract
+    // allows it); resolve it the same way SmrtCollection.get() would.
+    const resolved =
+      typeof intercepted === 'string'
+        ? resolveGetStringFilter(intercepted)
+        : intercepted;
+    return keysToSnakeCase(resolved as Record<string, unknown>);
+  }
```
[the remark this answers is not in the document]
Comment (willgriffin) [reply packages/core/src/object.ts:1683]: Confirmed — the finding is accurate, verified at source (`toSnakeCase()` in `packages/core/src/utils/naming.ts` strips a leading underscore via `.replace(/^_/, '')`, so `keysToSnakeCase()` in `interceptGetFilter()` would corrupt `_meta_type`/`_meta_data` if they appeared in interceptor-returned keys).

However, it's not reachable today: the only `beforeGet` interceptor registered anywhere in this repo is tenancy's (`packages/tenancy/src/interceptor.ts:294`), and it returns only `{ id, tenantId }` — no underscore-prefixed keys. The underscore-stripping behavior in `toSnakeCase` also predates this PR; #2365 only makes the path reachable from filters, it doesn't introduce the defect.

Tracked as https://github.com/happyvertical/smrt/issues/2417 for a proper fix (with a regression test, and a check of `keysToSnakeCase()`'s other callers) rather than blocking this PR's cross-tenant `getId()` fix behind an unreachable edge case.
